src/bootstrap/lib.rs

```diff
@@ -557,17 +557,26 @@ impl Build {
                 continue
             }
 
+            // `submodule.path` is the relative path to a submodule (from the repository root)
+            // `submodule_path` is the path to a submodule from the cwd
+
+            // use `submodule.path` when e.g. executing a submodule specific command from the
+            // repository root
+            // use `submodule_path` when e.g. executing a normal git command for the submodule
+            // (set via `currnet_dir`)
+            let submodule_path = self.src.join(submodule.path);
+
             if !submodule.path.exists() {
-                t!(fs::create_dir_all(&submodule.path));
+                t!(fs::create_dir_all(&submodule_path));
```

Comment:
Why do we need this?

* `State::NotInitialized` does not need this
* `State::OutOfSync` should have git create the directory when updating the submodule
* `State::MaybeDirty` should not need this (I think): If the submodule is dirty, it already exists. Also the code would be ineffectual, since it creates just the directory and does not set up the `.git` metadata